### PR TITLE
jsonsave function for sol

### PR DESCRIPTION
**Pull request by Alexander-Da-Great**
This branch implements json saving for solutions. This is just the implementation for solutions - excluding the diffs discussion for now. Json's produced via this method have a depth of two. It might be useful to have a hierarchical structure in the json sol that reflects the structure of the model a solution is formed from.

### CODE REVIEW

**Comment by bqpd** [comment]
whitelist this

**Comment by bqpd** [comment]
test this please

File: gpkit/solution_array.py [hunk 1]
```diff
@@ -592,6 +593,23 @@ def savetxt(self, filename="solution.txt", *, printmodel=True, **kwargs):
...
             f.write(self.table(**kwargs))
 
+    def savejson(self, filename="solution.json"):
+        "Saves solution table as a json file"
+        sol_dict = {}
+        # get list of variables
+        data = self["variables"]
+        # add appropriate data for each variable to the dictionary
+        for i, (k, v) in enumerate(data.items()):
+            key = str(k.name)
```
Comment (bqpd) [inline gpkit/solution_array.py:603]: This should probably be `str(k)` within a `for key in self.name_collision_varkeys():  key.descr["necessarylineage"] = True/False` environment, as in `def tables(...)` below.

**Review by bqpd — CHANGES_REQUESTED** [review]
made one note, otherwise looks great!

File: gpkit/solution_array.py [hunk 2]
```diff
@@ -592,6 +593,25 @@ def savetxt(self, filename="solution.txt", *, printmodel=True, **kwargs):
                 f.write(self.modelstr + "\n")
             f.write(self.table(**kwargs))
 
+    def savejson(self, filename="solution.json"):
```
Comment (bqpd) [inline gpkit/solution_array.py:596]: Looks great! There should be a test that just verifies the output of this on e.g. the Peformance Modeling example...will add it when I can or you're welcome to take a stab!

Comment (Alexander-Da-Great) [reply gpkit/solution_array.py:596]: Test is up under the performance modeling example in t_examples.py. Let me know if this is the kind of test you would be looking for - my stab may have been more like a gouge.

Comment (bqpd) [inline gpkit/solution_array.py:596]: do you want to add an argument to turn only certain variables into json?

File: gpkit/solution_array.py [hunk 3]
```diff
@@ -592,6 +593,25 @@ def savetxt(self, filename="solution.txt", *, printmodel=True, **kwargs):
...
+        "Saves solution table as a json file"
+        sol_dict = {}
+        # add appropriate data for each variable to the dictionary
+        for key in self.name_collision_varkeys():
+            key.descr["necessarylineage"] = True
+        data = self["variables"]
+        for i, (k, v) in enumerate(data.items()):
+            key = str(k)
+            if isinstance(v, np.ndarray):
+                val = {'v': v.tolist(), 'u': str(k.unitstr())}
```
Comment (bqpd) [inline gpkit/solution_array.py:606]: GPkit uses `"` double quotes, sorry for the bother!

Comment (bqpd) [inline gpkit/solution_array.py:606]: is there a reason that just `k.unitstr()` doesn't work here?

File: gpkit/solution_array.py [hunk 4]
```diff
@@ -592,6 +593,25 @@ def savetxt(self, filename="solution.txt", *, printmodel=True, **kwargs):
...
+            else:
+                val = {'v': v, 'u': str(k.unitstr())}
+            sol_dict[key] = val
+        for key in self.name_collision_varkeys():
+            del key.descr["necessarylineage"]
+        with open(filename, "w") as f:
+            json.dump(sol_dict, f)
```
Comment (bqpd) [inline gpkit/solution_array.py:613]: should this also return the json as a string, in case the user wants it? (i.e. `return json.dumps(sol_dict)`)

maybe if the filename is `None` it doesn't save and only does that.

File: gpkit/tests/t_examples.py [hunk 5]
```diff
@@ -192,6 +193,17 @@ def test_performance_modeling(self, example):
...
+        json_dict = {}
+        with open("solution.json", "r") as rf:
+            json_dict = json.load(rf)
+        for var in sol["variables"]:
+            self.assertTrue(np.all(json_dict[str(var.key)]['v']
+                                   == sol["variables"][var.key]))
+            self.assertEqual(json_dict[str(var.key)]['u'], var.unitstr())
+
+
+
```
Comment (bqpd) [inline gpkit/tests/t_examples.py:206]: not sure how these extra newlines crept in but otherwise this test looks great!

File: gpkit/tests/t_examples.py [hunk 6]
```diff
@@ -177,7 +178,7 @@ def test_model_var_access(self, example):
 
     def test_performance_modeling(self, example):
         m = Model(example.M.cost, Loose(example.M), example.M.substitutions)
-
+        NamedVariables.reset_modelnumbers()
```
Comment (bqpd) [inline gpkit/tests/t_examples.py:181]: why is this necessary? (may be worth a comment indicating)

**Review by bqpd — CHANGES_REQUESTED** [review]
It looks feature-complete! Added notes on a couple speculative features and lint.

File: gpkit/solution_array.py (hunk 2 above)
Comment (Alexander-Da-Great) [reply gpkit/solution_array.py:596]: This would be a great addition. I was talking to Philippe about how it would be nice to have solution jsons organized with a hierarchical structure according to sub-models. It could be nice to have the option to print only variables in a certain sub-model rather than imposing a top down structure on the json.

File: gpkit/tests/t_examples.py (hunk 6 above)
Comment (Alexander-Da-Great) [reply gpkit/tests/t_examples.py:181]: When developing I encountered an issue where mode1.var != model.var - that has since disappeared. Not sure how this got resolved, either way I will removed the   NamedVariables line.

File: gpkit/solution_array.py (hunk 2 above)
Comment (bqpd) [reply gpkit/solution_array.py:596]: `showvars = self._parse_showvars(showvars)` cleans up some of the various input types, then just intersect / check for inclusion of "k" in there. `parse_showvars` should handle vectorvariables for you, but worth a check if you end up implementing this!

Comment (bqpd) [reply gpkit/solution_array.py:596]: looks great to me

**Comment by bqpd** [comment]
@Alexander-Da-Great, I merged the recent changes to master so that the unit tests would run.

File: gpkit/solution_array.py (hunk 4 above)
Comment (Alexander-Da-Great) [reply gpkit/solution_array.py:613]: created a boolean called printjson that will return the string if flagged true

File: gpkit/solution_array.py (hunk 2 above)
Comment (Alexander-Da-Great) [reply gpkit/solution_array.py:596]: Would love to hear any input on what would make sense in the "printing variables of specific sub-models" department

Comment (bqpd) [reply gpkit/solution_array.py:596]: with this method you can thankfully just use the models' `varkeys` attribute, as in this section from the performance modeling example:

```python
vars_of_interest = set(AC.varkeys)
# note that there's two ways to access submodels
assert (MISSION["flight segment"]["aircraft performance"]
        is MISSION.fs.aircraftp)
vars_of_interest.update(MISSION.fs.aircraftp.unique_varkeys)
vars_of_interest.add(M["D"])
print(sol.summary(vars_of_interest))
print(sol.table(tables=["loose constraints"]))
```

File: gpkit/solution_array.py (hunk 4 above)
Comment (bqpd) [reply gpkit/solution_array.py:613]: hmm maybe `returnjson`?

File: gpkit/solution_array.py (hunk 2 above)
Comment (Alexander-Da-Great) [reply gpkit/solution_array.py]: Thanks for the input. Added the showvars parameter to the savejson for applying this logic.

**Comment by bqpd** [comment]
test this please

File: gpkit/solution_array.py [hunk 7]
```diff
@@ -592,6 +593,31 @@ def savetxt(self, filename="solution.txt", *, printmodel=True, **kwargs):
...
+            key = str(k)
+            if isinstance(v, np.ndarray):
+                val = {"v": v.tolist(), "u": k.unitstr()}
+            else:
+                val = {"v": v, "u": k.unitstr()}
+            sol_dict[key] = val
+        for key in self.name_collision_varkeys():
+            del key.descr["necessarylineage"]
+        if printjson:
+            return str(sol_dict)
```
Comment (bqpd) [inline gpkit/solution_array.py:616]: shouldn't this be the JSON string (via `json.dumps`?)

Comment (bqpd) [reply gpkit/solution_array.py:616]: If you don't think this feature will be used, feel free to remove it!

Comment (bqpd) [reply gpkit/solution_array.py:616]: Other than this q, this looks perfect.

**Comment by bqpd** [comment]
@Alexander-Da-Great sorry for the backseat driving with `printjson` :grimacing: I was thinking it'd be useful to have a "canonical" json representation that various streaming APIs could use, but realized that any early API will want to be idiosyncratic with its messaging (until a good standard is found).

**Review by bqpd — APPROVED** [review]

**Comment by bqpd** [comment]
test this please
